File: box-and-box/ports/python/supervise.py
"""supervise.py - trajectory supervision (v0.4).

SAFETY specs (G not bad) extend the alethic floor across time - a runtime SHIELD.
LIVENESS specs (F goal, GF progress) extend the deontic OUGHT across time - a
temporal obligation that, unmet at the horizon, triggers contrary-to-duty
escalation. Mirrors supervise.mjs. Laws TB1-TB3.
"""

from temporal import progress, monitor, show


def TemporalSpec(id, formula, kind="safety", ctd=None):
    return {"id": id, "formula": formula, "kind": kind, "ctd": ctd}


def _first_vio(online):
    try:
        return online.index("vio")
    except ValueError:
        return None


def supervise(trajectory, specs):
    reports = []
    for spec in specs:
        m = monitor(spec["formula"], trajectory)
        r = {
            "id": spec["id"], "kind": spec["kind"], "formula": show(spec["formula"]),
            "verdict": m["verdict"], "online": m["online"], "decidedAt": m["decidedAt"],
            "violatedAt": _first_vio(m["online"]) if spec["kind"] == "safety" else None,
        }
        if spec["kind"] == "liveness" and m["verdict"] == "violated":
            r["escalation"] = spec.get("ctd") or "escalate-to-human"
            r["reason"] = (f"liveness obligation {show(spec['formula'])} unmet within horizon "
                           f"({len(trajectory)} steps)")
        reports.append(r)

    safety_violated = [r for r in reports if r["kind"] == "safety" and r["verdict"] == "violated"]
    liveness_unmet = [r for r in reports if r["kind"] == "liveness" and r["verdict"] == "violated"]
    return {
        "reports": reports,
        "safe": len(safety_violated) == 0,
        "escalation": ({"required": True,
                        "specs": [{"id": r["id"], "repair": r["escalation"], "reason": r["reason"]} for r in liveness_unmet]}
                       if liveness_unmet else None),
        "note": _note(safety_violated, liveness_unmet, len(trajectory)),
    }


def residual_of(formula, history):
    """the one-step shield: residual of a safety spec after a history of states."""
    f = formula
    for s in history:
        f = progress(f, s)
    return f


def guard(residual, next_state):
    return progress(residual, next_state)["t"] == "false"


def _note(safety, liveness, n):
    if safety:
        return (f"UNSAFE - \u201c{safety[0]['id']}\u201d violated at step {safety[0]['violatedAt']}; "
                f"the safety shield would have pruned that transition.")
    if liveness:
        return (f"safe, but a liveness obligation went unmet at the horizon ({n} steps) -> escalation: "
                + ", ".join(r["escalation"] for r in liveness) + ".")
    return f"all specs satisfied over {n} steps - safe and live."

File: box-and-box/ports/go/supervise.go
package main

// supervise.go — trajectory supervision (port of supervise.mjs). Laws TB1–TB3.

// TemporalSpec couples a formula with a kind (safety/liveness) and CTD repair.
type TemporalSpec struct {
	ID      string
	Formula *Formula
	Kind    string // "safety" | "liveness"
	Ctd     string
}

// NewTemporalSpec builds a TemporalSpec with kind default "safety".
func NewTemporalSpec(id string, formula *Formula, kind, ctd string) TemporalSpec {
	if kind == "" {
		kind = "safety"
	}
	return TemporalSpec{ID: id, Formula: formula, Kind: kind, Ctd: ctd}
}

func firstVio(online []string) int {
	for i, v := range online {
		if v == "vio" {
			return i
		}
	}
	return -1
}

// SuperviseReport is a per-spec report.
type SuperviseReport struct {
	ID         string
	Kind       string
	Verdict    string
	Online     []string
	DecidedAt  int
	ViolatedAt int // -1 if none / not safety
	Escalation string
	Reason     string
}

// SuperviseEscalation aggregates unmet liveness obligations.
type SuperviseEscalation struct {
	Required bool
	Specs    []EscSpec
}

// EscSpec is one escalated spec.
type EscSpec struct {
	ID     string
	Repair string
	Reason string
}

// SuperviseResult is the output of supervise.
type SuperviseResult struct {
	Reports    []SuperviseReport
	Safe       bool
	Escalation *SuperviseEscalation
}

func supervise(trajectory []State, specs []TemporalSpec) SuperviseResult {
	reports := []SuperviseReport{}
	for _, spec := range specs {
		m := monitor(spec.Formula, trajectory)
		r := SuperviseReport{
			ID: spec.ID, Kind: spec.Kind, Verdict: m.Verdict, Online: m.Online,
			DecidedAt: m.DecidedAt, ViolatedAt: -1,
		}
		if spec.Kind == "safety" {
			r.ViolatedAt = firstVio(m.Online)
		}
		if spec.Kind == "liveness" && m.Verdict == "violated" {
			if spec.Ctd != "" {
				r.Escalation = spec.Ctd
			} else {
				r.Escalation = "escalate-to-human"
			}
			r.Reason = "liveness obligation unmet within horizon"
		}
		reports = append(reports, r)
	}
	safetyViolated := 0
	var livenessUnmet []SuperviseReport
	for _, r := range reports {
		if r.Kind == "safety" && r.Verdict == "violated" {
			safetyViolated++
		}
		if r.Kind == "liveness" && r.Verdict == "violated" {
			livenessUnmet = append(livenessUnmet, r)
		}
	}
	res := SuperviseResult{Reports: reports, Safe: safetyViolated == 0}
	if len(livenessUnmet) > 0 {
		esc := &SuperviseEscalation{Required: true}
		for _, r := range livenessUnmet {
			esc.Specs = append(esc.Specs, EscSpec{ID: r.ID, Repair: r.Escalation, Reason: r.Reason})
		}
		res.Escalation = esc
	}
	return res
}

// residualOf : fold progress over a history.
func residualOf(formula *Formula, history []State) *Formula {
	f := formula
	for _, s := range history {
		f = progress(f, s)
	}
	return f
}

// guard : would the next state drive the residual to ⊥?
func guard(residual *Formula, nextState State) bool {
	return progress(residual, nextState).T == "false"
}
